### PR TITLE
V22F-507 Code coverage IndicatorConfig

### PR DESCRIPTION
**Pull request by larsdekwant**
Created tests to reach 100% code coverage on the IndicatorConfig. 
Note that the loaded Font is not asserted right now, because of the way the font is currently loaded. Also the Font class currently does not have any members to represent a name and/or font-size, thus this is not possible to assert.
Could be implemented later when the font functionality is extended.

### CODE REVIEW

**Review by shoekzema — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp [hunk 1]
```diff
@@ -27,17 +27,29 @@ class Random
         return NextFloat(p_max - p_min) + p_min;
     }
 
+    /// @return A random float with a max value (included)
+    float NextFloatIncl(float p_max)
+    {
+        return static_cast<float>(NextUShort()) / (65535.0f / p_max);
```
[the remark this answers is not in the document]
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp:33]: is not a magic number and is from the Random file

File: speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp [hunk 2]
```diff
@@ -249,6 +250,11 @@ static void SaveSettings(void* /* dummy */)
     mediator->SetPControlSettings(m_pControl);
     mediator->SetBlackBoxFilePath(m_blackBoxFilePath);
 
+    // Load intervention indicator texture from XML file (unchecked max p_path size: 256)
+    char path[PATH_BUF_SIZE];
+    snprintf(path, PATH_BUF_SIZE, CONFIG_XML_DIR_FORMAT, GfDataDir());
+    IndicatorConfig::GetInstance()->LoadIndicatorData(path);
```
[the remark this answers is not in the document]
Comment (larsdekwant) [reply speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/ResearcherMenu.cpp]: Moved to `Mediator::RaceStart`

**Review by shoekzema — APPROVED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp [hunk 3]
```diff
@@ -9,10 +9,22 @@
 
 #include "IndicatorConfig.h"
 
-/// @brief Initialize the indicator configuration.
-void IndicatorConfig::Initialize()
+/// @brief Loads the indicator data of every intervention action from config file in the given path
+void IndicatorConfig::LoadIndicatorData(const char* p_path)
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp:13]: Missing @param

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp [hunk 4]
```diff
@@ -1,7 +1,351 @@
...
+/// @brief            Creates randomly generated screen positions
+/// @param p_rnd      The random generator reference to use
+/// @param p_validity The validity flags to use when generating data
+/// @return           The generated screen position
+tScreenPosition CreateRandomScreenPosition(Random& p_rnd, DataGeneration p_gen)
+{
+    if (p_gen & INVALID_SCR_POS)
+    {
+        return {p_rnd.NextFloat(1.0f, FLT_MAX), p_rnd.NextFloat(1.0f, FLT_MAX)};
+    }
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp:62]: Should  this also test negative values?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp [hunk 5]
```diff
@@ -1,7 +1,351 @@
...
+/// @brief Creates a random indicator data object, writes this to an xml file,
+///        then reads in this xml file using IndicatorConfig::LoadIndicatorData
+///        Asserts whether the loaded in data is equal to the randomly generated data.
+TEST(IndicatorConfigTests, LoadIndicatorDataFromXML)
+{
+    // Gf module contains all the XML reading/writing functions.
+    GfInit(GF_LOGGING_DISABLE);
+    GfSetDataDir(SD_DATADIR_SRC);
+    ASSERT_TRUE(SetupSingletonsFolder());
+    Random rnd;
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp:258]: Would it be possible to log the seed that is being used so it's easier to replicate a fail?

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp [hunk 6]
```diff
@@ -5,20 +5,35 @@
 #define BYTE   uint8_t
 #define USHORT uint16_t
 
+#define SHORT_MAXIMUM_VALUE 65535
+#define SHORT_MAX_VALUE     65536
```
Comment (suppergerrie2) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp:9]: Any reason to not use ` std::numeric_limits::max()`?

**Review by suppergerrie2 — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp (hunk 5 above)
Comment (larsdekwant) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontendUnitTests/IndicatorConfigTests.cpp:266]: I will look into this using `SetUp` and `TearDown` functionality of GTest. I think it should be possible to log the random seed used in every test.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp (hunk 6 above)
Comment (larsdekwant) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp:9]: This was refactored by @KungFuDonkey, I will replace it with `USHRT_MAX`

Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/rppUtils/Random.hpp:9]: no reason to use a function where a constant can used

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp [hunk 7]
```diff
@@ -167,18 +160,17 @@ IndicatorConfig* IndicatorConfig::GetInstance()
...
-    m_instance = (IndicatorConfig*)pointerValue;
-    return m_instance;
+    // File exists -> read pointer                      // Only hit when instances are created in multiple modules, which is impossible in the testing environment.
+    std::string pointerName("00000000");                // @NOCOVERAGE
+    std::ifstream file(filepath);                       // @NOCOVERAGE
+    getline(file, pointerName);                         // @NOCOVERAGE
+    file.close();                                       // @NOCOVERAGE
+    int pointerValue = stoi(pointerName, nullptr, 16);  // @NOCOVERAGE
+    m_instance = (IndicatorConfig*)pointerValue;        // @NOCOVERAGE
+    return m_instance;                                  // @NOCOVERAGE
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/frontend/IndicatorConfig.cpp:173]: It is possible to hit these lines PR #98 has a way of doing this. Maybe this can be fixed when the PR is either merged into main, or if you copy its way of dealing with this.

**Review by KungFuDonkey — CHANGES_REQUESTED** [review]

**Comment by larsdekwant** [comment]
Also had to fix DecisionTests because I separated the loading of the `IndicatorConfig` from the `GetInstance()` for testability and a more logical program initialisation. When the `IndicatorConfig` is used for testing, it should thus also be loaded with data beforehand.

Also refactored the MediatorTests::ReadFromFile in consultation with @KungFuDonkey . This test had redundant code and required too much `Test` specific code in the actual source code.

**Review by KungFuDonkey — APPROVED** [review]

**Review by suppergerrie2 — APPROVED** [review]
